rfcs/2022-08-25-12217-handling-discarded-events.md

```diff
@@ -0,0 +1,369 @@
+2. Update configuration validation to require that all outputs are handled.
+3. Add a new configuration for discarding or rejecting outputs.
+
+### User Experience
+
+#### Add Discarded Event Output
+
+We will introduce a new output to all components that would otherwise discard events, named
+`discards`.
+
+#### Validate Handling All Outputs
+
+Additionally, we will enhance the configuration validation to determine if all outputs are
+handled. This will provide notification of the new discard outputs and ensure users are aware of the
+new feature.
+
+In order to avoid breaking existing configurations, this validation will initially produce a
+deprecation warning, notifying the user that an output is present but unhandled. In a later version,
+this deprecation will be turned into a hard error, preventing the acceptance of a configuration with
+unhandled outputs.
+
+We will also add a new command-line option and environment variable to opt into the stricter
+validation for users that want the additional assurance this provides.
+
+#### Simplify Discarded Output Handling
+
+This enforced handling will create considerable complication for most users, as well as the
+potential of increased overhead and reduced performance if extra components have to be configured
+just to handle the outputs that now need to be connected somewhere. To reduce this overhead, we will
+add a new optional configuration setting to indicate the internal disposition of each output:
+`outputs.NAME.disposition`. This will have two non-default values, `"drop"` to mark all events going
+to that output as delivered and then drop them, and `"reject"` to mark them as having failed.  If
+this setting is not present and a component has an unhandled output, a default behavior of `"drop"`
+will be assumed, which matches the current behavior.
+
+So, to configure the new discard output as discarding events with an error result, users would add the
+following:
+
+```toml
+[transforms.example]
+type = "remap"
+…
+outputs.discards.disposition = "reject"
+```
+
+### Implementation
+
+#### Configuration
+
+The new options will show up in all of the enclosing configuration option structures:
+
+```rust
+struct SourceOuter {
+    …
+
+    /// The configuration of all named outputs.
+    outputs: HashMap<String, OutputConfig>,
+}
+
+struct TransformOuter {
+    …
+
+    /// The configuration of all named outputs.
+    outputs: HashMap<String, OutputConfig>,
+}
+
+struct SinkOuter {
+    …
+
+    /// The configuration of all named outputs.
+    outputs: HashMap<String, OutputConfig>,
+}
+
+struct OutputConfig {
+    #[serde(default)]
+    disposition: OutputDisposition,
+}
+
+enum OutputDisposition {
+    /// Default behavior, send the event on to further components that name this as an input.
+    Send,
+    /// Discard the event and mark it is as delivered.
+    Drop,
+    /// Discard the event and mark it as failed.
+    Reject,
+}
+```
+
+#### Sources
```

Comment:
Which of the possible discard scenarios apply to sources? It's not obvious that any do from the list above.

Reply:
Hmm, I missed that. I was fairly certain some could error or otherwise discard after creating events (other than on output disconnect), but I'll have to review the code to see that.

Reply:
Okay, AFAICT there is _exactly one_ source that can discard events between creating them and sending them to the output: `kubernetes_logs` loads events from a file source, creates events, and then runs a transform on them to parse data out, which may discard the created events if the contents don't parse properly. So, I *think* this still applies to sources because of that.